ggml/src/ggml-vulkan/vulkan-shaders/mul_mat_vecq.comp

```diff
@@ -0,0 +1,201 @@
+#define INVOCATION_ID gl_LocalInvocationID.x
+#endif
+
+#define MMQ
+#define B_TYPE block_q8_1_packed32
+
+#include "mul_mat_vec_base.comp"
+
+layout(local_size_x_id = 0, local_size_y = 1, local_size_z = 1) in;
+
+#define K_PER_ITER 32
+
+const uint GROUP_SIZE = 8;
+const uint GROUPS_PER_WARP = (BLOCK_SIZE / GROUP_SIZE);
+
+#include "mul_mmq_funcs.comp"
+
+uint a_offset, b_offset, d_offset, y_offset;
+
+#ifdef USE_SUBGROUPS
+void reduce_result_grouped(inout FLOAT_TYPE temp[NUM_COLS][NUM_ROWS], const in uint32_t d_offset, const in uint32_t first_row, const in uint32_t num_rows, const in uint32_t tid_in_group) {
+    [[unroll]] for (uint j = 0; j < NUM_COLS; ++j) {
+        [[unroll]] for (uint n = 0; n < num_rows; ++n) {
+            temp[j][n] = subgroupClusteredAdd(temp[j][n], GROUP_SIZE);
+        }
+    }
+
+    if (tid_in_group == 0) {
+        [[unroll]] for (uint j = 0; j < NUM_COLS; ++j) {
+            [[unroll]] for (uint n = 0; n < num_rows; ++n) {
+                data_d[j*p.batch_stride_d + d_offset + first_row + n] = D_TYPE(temp[j][n]);
+            }
+        }
+    }
+}
+#else
+void reduce_result_grouped(const in FLOAT_TYPE temp[NUM_COLS][NUM_ROWS], const in uint32_t d_offset, const in uint32_t first_row, const in uint32_t num_rows, const in uint32_t tid_in_group) {
+    const uint tid = INVOCATION_ID;
+    // sum up partial sums and write back result
+    [[unroll]] for (uint j = 0; j < NUM_COLS; ++j) {
+        [[unroll]] for (uint n = 0; n < num_rows; ++n) {
+            tmpsh[j][n][tid] = temp[j][n];
+        }
+    }
+    barrier();
+    if (tid_in_group < 4) {
+        [[unroll]] for (uint j = 0; j < NUM_COLS; ++j) {
+            [[unroll]] for (uint n = 0; n < num_rows; ++n) {
+                tmpsh[j][n][tid] += tmpsh[j][n][tid + 4];
+            }
+        }
+    }
+    barrier();
+    if (tid_in_group < 2) {
+        [[unroll]] for (uint j = 0; j < NUM_COLS; ++j) {
+            [[unroll]] for (uint n = 0; n < num_rows; ++n) {
+                tmpsh[j][n][tid] += tmpsh[j][n][tid + 2];
+            }
+        }
+    }
+    barrier();
+    if (tid_in_group == 0) {
+        [[unroll]] for (uint j = 0; j < NUM_COLS; ++j) {
+            [[unroll]] for (uint n = 0; n < num_rows; ++n) {
+                data_d[j*p.batch_stride_d + d_offset + first_row + n] = D_TYPE(tmpsh[j][n][tid] + tmpsh[j][n][tid + 1]);
+            }
+        }
+    }
+}
+#endif
+
+int32_t cache_b_qs[8];
+vec2 cache_b_ds;
+
+void iter(inout FLOAT_TYPE temp[NUM_COLS][NUM_ROWS], const uint first_row, const uint num_rows, const uint tid_in_group, const uint i) {
+    [[unroll]] for (uint j = 0; j < NUM_COLS; ++j) {
+        const uint col = i*GROUP_SIZE + K_PER_ITER*tid_in_group;
+
+        // Preload data_b block
+        const uint b_block_idx = (j*p.batch_stride_b + col) / QUANT_K_Q8_1 + b_offset;
+        cache_b_ds = vec2(data_b[b_block_idx].ds);
+        [[unroll]] for (uint k = 0; k < 8; k++) {
+            cache_b_qs[k] = data_b[b_block_idx].qs[k];
```

Comment:
You need a barrier after these shared memory stores, and either after the loads or before the stores for the next iteration.

Seems like you can cut down the loads by having the first 8 threads each do one of the iterations. And ds could just go straight to registers rather than the extra copy through shared memory.

Reply:
That's not shared memory

Reply:
Oops. Maybe it's worth loading the qs values through shared memory? If the issue is with too many small loads like you suggested, then copying through shared memory ought to help.

Reply:
Actually, I guess I can't tell if the b_block_idx value is shared between threads. So maybe this idea doesn't work.

Reply:
Another idea might be to add padding to the q8_1 struct so you can do 16B loads rather than 4B loads.

Reply:
Yeah, that might be worth it. I know the cuda backend stacks 4 q8_1 blocks in a struct for that reason.